bundler/lib/bundler/checksum.rb

```diff
@@ -228,16 +223,32 @@ def to_lock(spec)
       private
 
       def register_checksum(lock_name, checksum)
-        return unless checksum
-        checksums = (store[lock_name] ||= {})
-        existing = checksums[checksum.algo]
+        existing = fetch_checksum(lock_name, checksum.algo)
 
-        if !existing
-          checksums[checksum.algo] = checksum
-        elsif existing.merge!(checksum)
-          checksum
+        if existing
+          merge_checksum(lock_name, checksum, existing)
         else
-          raise ChecksumMismatchError.new(lock_name, existing, checksum)
+          store_checksum(lock_name, checksum)
+        end
+      end
+
+      def merge_checksum(lock_name, checksum, existing)
+        existing.merge!(checksum) || raise(ChecksumMismatchError.new(lock_name, existing, checksum))
+      end
+
+      def store_checksum(lock_name, checksum)
+        @store_mutex.synchronize do
+          (@store[lock_name] ||= {}).store(checksum.algo, checksum)
+        end
+      end
+
+      def fetch_checksum(lock_name, algo)
+        fetch_checksums(lock_name)&.fetch(algo, nil)
```

Comment:
@segiddins This is a fetch on a Hash (the nested one). We're syncing other reads, but is this one safe? Note the second half of this method is not synced.

Reply:
any read or write on data that is (at the time of access) shared across threads (or fibers, actors, etc), needs to be synchronized. typically, this means not building upon primitive access methods directly, but rather exposing a set of public methods that each synchronizes access to the shared data